dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchMemoryRecordService.cs

```diff
@@ -0,0 +1,357 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.Json.Nodes;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Search.Documents;
+using Azure.Search.Documents.Indexes;
+using Azure.Search.Documents.Models;
+using Microsoft.SemanticKernel.Memory;
+
+namespace Microsoft.SemanticKernel.Connectors.AzureAISearch;
+
+/// <summary>
+/// Service for storing and retrieving memory records, that uses Azure AI Search as the underlying storage.
+/// </summary>
+/// <typeparam name="TDataModel">The data model to use for adding, updating and retrieving data from storage.</typeparam>
+public sealed class AzureAISearchMemoryRecordService<TDataModel> : IMemoryRecordService<string, TDataModel>
+    where TDataModel : class
+{
+    /// <summary>A set of types that a key on the provided model may have.</summary>
+    private static readonly HashSet<Type> s_supportedKeyTypes = new()
+    {
+        typeof(string)
+    };
+
+    /// <summary>A set of types that vectors on the provided model may have.</summary>
+    private static readonly HashSet<Type> s_supportedVectorTypes = new()
+    {
+        typeof(ReadOnlyMemory<float>),
+        typeof(ReadOnlyMemory<float>?)
```

Comment:
Yeah, double isn't currently supported, and converting from double to float without the developer knowing is not great, since they may lose precision. My understanding is that someone should always try to use the model to generate the exact level of precision needed and not post process it after to fit the specific storage.
There are a bunch of new field types in preview actually but at the moment the Client SDK only supports creating float32 fields for vectors.  It's actually hardcoded on field definition, so we are constrained anyway.
https://learn.microsoft.com/en-us/rest/api/searchservice/supported-data-types#edm-data-types-for-vector-fields

I'd rather start with these restrictions and loosen them if needed later, since that would be easy, but tightening them later would be impossible.  That does mean more maintenance though as services churn, but it would be pretty small changes, and just making sure our mappings still work with the new types.